### PR TITLE
Port TemplateBinder to dispatcher

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

File: src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs [hunk 1]
```diff
@@ -41,320 +33,29 @@ public class TemplateBinder
...
-            _defaults = defaults;
-
-            // Any default that doesn't have a corresponding parameter is a 'filter' and if a value
-            // is provided for that 'filter' it must match the value in defaults.
-            _filters = new RouteValueDictionary(_defaults);
-            foreach (var parameter in _template.Parameters)
-            {
-                _filters.Remove(parameter.Name);
-            }
+            _binder = new RoutePatternBinder(urlEncoder, pool, template.ToRoutePattern(), defaults);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs]: /cc @eilon - let's talk about this one

File: src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs [hunk 2]
```diff
@@ -21,10 +22,6 @@ namespace Microsoft.AspNetCore.Routing.Tree
     /// </summary>
     public class TreeRouter : IRouter
     {
-        // Key used by routing and action selection to match an attribute route entry to a
-        // group of action descriptors.
-        public static readonly string RouteGroupKey = "!__route_group";
-
```
Comment (jbagga) [inline src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs]: Why is this safe to delete?

Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs]: OH! - I deleted it from the wrong spot. Will fix.

Comment (jbagga) [reply src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs]: It wasn't being used in TreeRouter anyway though. Perhaps it can stay deleted

Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs]: it's public in a public class, it needs to stay.

File: src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs (hunk 1 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs]: Discussed in person, @Eilon doesn't have concerns about this.

**Pull request merged**